### DISCUSSION TITLE
Very slow association with invalid adresses

### DISCUSSION THREAD

**Discussion by qarmin** [Enet4/dicom-rs#480] [Q&A]
I have this code
```
    debug!("Establishing association with '{}'...", &addr);
    let mut scu_opt = ClientAssociationOptions::new()
        .with_abstract_syntax(abstract_syntax)
        .called_ae_title(aetitle)
        .max_pdu_length(max_pdu_length);

    if !station_name.is_empty() {
        scu_opt = scu_opt.calling_ae_title(station_name);
    }

    let scu = scu_opt
        .establish_with(addr)
        .map_err(|_e| DicomError::CommonError(format!("Failed to establish connection with {addr}")))?;

    debug!("Association established");
    Ok(scu)
}
```
and after trying to connect to invalid address, after 2 minutes I have info, that dicom failed to establish connection.
Can this be speed up? 
Can I use any workaround?
```
11:29:01.681 [DEBUG] Establishing association with '192.168.10.124:4242'...
11:31:16.409 [ERROR] Failed to get patients from worklist: Failed to establish connection with 192.168.10.124:4242
```

**Comment by qarmin** [comment]
Looks that problem is with this function - https://github.com/Enet4/dicom-rs/blob/7c0e5ab895e2f57c432cece41077f13abd4d7f71/ul/src/association/client.rs#L373

I see that there is also function connect_timeout in TcpStream - can this function be added here with time provided by user?

**Comment by Enet4** [comment]
There is no explicit timeout for connection attempts, so the decision to bail out is mostly delegated to the underlying system.

So I agree that the `dicom_ul` association API should provide a way to specify certain timeouts, at least one for connection establishment and another one for DIMSE responses.
